In [0]:
from google.colab import files
datafiles = files.upload()

Saving Data set -Churn modelling.xlsx to Data set -Churn modelling.xlsx


In [0]:
import pandas as pd

In [0]:
data = pd.read_excel('Data set -Churn modelling.xlsx', sheet_name=1)

In [0]:
data.shape

(3333, 11)

In [0]:
df = pd.DataFrame(data)

In [0]:
df.head()

Churn  AccountWeeks  ContractRenewal  DataPlan  DataUsage  CustServCalls  \
0      0           128                1         1        2.7              1   
1      0           107                1         1        3.7              1   
2      0           137                1         0        0.0              0   
3      0            84                0         0        0.0              2   
4      0            75                0         0        0.0              3   

   DayMins  DayCalls  MonthlyCharge  OverageFee  RoamMins  
0    265.1       110           89.0        9.87      10.0  
1    161.6       123           82.0        9.78      13.7  
2    243.4       114           52.0        6.06      12.2  
3    299.4        71           57.0        3.10       6.6  
4    166.7       113           41.0        7.42      10.1

In [0]:
df_copy = df.copy()

In [0]:
df['Churn'].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

In [0]:
df_copy.columns

Index(['Churn', 'AccountWeeks', 'ContractRenewal', 'DataPlan', 'DataUsage',
       'CustServCalls', 'DayMins', 'DayCalls', 'MonthlyCharge', 'OverageFee',
       'RoamMins'],
      dtype='object')

In [0]:
df_copy.isnull().sum()

Churn              0
AccountWeeks       0
ContractRenewal    0
DataPlan           0
DataUsage          0
CustServCalls      0
DayMins            0
DayCalls           0
MonthlyCharge      0
OverageFee         0
RoamMins           0
dtype: int64

In [0]:
df_1=df_copy.drop(['Churn','ContractRenewal','DataPlan'],axis=1)

In [0]:
df_1.shape

(3333, 8)

In [0]:
df_1.head()

AccountWeeks  DataUsage  CustServCalls  DayMins  DayCalls  MonthlyCharge  \
0           128        2.7              1    265.1       110           89.0   
1           107        3.7              1    161.6       123           82.0   
2           137        0.0              0    243.4       114           52.0   
3            84        0.0              2    299.4        71           57.0   
4            75        0.0              3    166.7       113           41.0   

   OverageFee  RoamMins  
0        9.87      10.0  
1        9.78      13.7  
2        6.06      12.2  
3        3.10       6.6  
4        7.42      10.1

In [0]:
col = df_1.columns

In [0]:
from sklearn import preprocessing

In [0]:
scaler = preprocessing.MinMaxScaler()

In [0]:
scaled_df = scaler.fit_transform(df_1)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/data.py:323: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


In [0]:
scaled_df = pd.DataFrame(scaled_df, columns=col)

In [0]:
scaled_df.head()

AccountWeeks  DataUsage  CustServCalls   DayMins  DayCalls  MonthlyCharge  \
0      0.524793   0.500000       0.111111  0.755701  0.666667       0.770812   
1      0.438017   0.685185       0.111111  0.460661  0.745455       0.698869   
2      0.561983   0.000000       0.000000  0.693843  0.690909       0.390545   
3      0.342975   0.000000       0.222222  0.853478  0.430303       0.441932   
4      0.305785   0.000000       0.333333  0.475200  0.684848       0.277492   

   OverageFee  RoamMins  
0    0.542606     0.500  
1    0.537658     0.685  
2    0.333150     0.610  
3    0.170423     0.330  
4    0.407916     0.505

In [0]:
cat = df.loc[:,['ContractRenewal', 'DataPlan']]

In [0]:
churn = df['Churn']

In [0]:
result = pd.concat([scaled_df,cat],axis=1)

In [0]:
result.shape

(3333, 10)

In [0]:
result.head()

AccountWeeks  DataUsage  CustServCalls   DayMins  DayCalls  MonthlyCharge  \
0      0.524793   0.500000       0.111111  0.755701  0.666667       0.770812   
1      0.438017   0.685185       0.111111  0.460661  0.745455       0.698869   
2      0.561983   0.000000       0.000000  0.693843  0.690909       0.390545   
3      0.342975   0.000000       0.222222  0.853478  0.430303       0.441932   
4      0.305785   0.000000       0.333333  0.475200  0.684848       0.277492   

   OverageFee  RoamMins  ContractRenewal  DataPlan  
0    0.542606     0.500                1         1  
1    0.537658     0.685                1         1  
2    0.333150     0.610                1         0  
3    0.170423     0.330                0         0  
4    0.407916     0.505                0         0

In [0]:
from imblearn.over_sampling import SMOTE


In [0]:
smt = SMOTE()

In [0]:
from sklearn.model_selection import train_test_split

In [0]:
X = result

In [0]:
cols = result.columns

In [0]:
X.shape

(3333, 10)

In [0]:
y = df['Churn']

In [0]:

y.shape

(3333,)

In [0]:
X.shape

(3333, 10)

In [0]:
X, y =smt.fit_sample(X, y)

In [0]:
from sklearn.linear_model import LogisticRegression

In [0]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size= 0.3, random_state= 0)

In [0]:

logreg = LogisticRegression()
logreg.fit(X_train, Y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [0]:
y_pred= logreg.predict(X_test)

In [0]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(Y_test, y_pred)
print(confusion_matrix)

[[666 183]
 [202 659]]


In [0]:
from sklearn.metrics import classification_report
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.78      0.78       849
           1       0.78      0.77      0.77       861

   micro avg       0.77      0.77      0.77      1710
   macro avg       0.77      0.77      0.77      1710
weighted avg       0.78      0.77      0.77      1710



For Buisness the person churning is more important and here we see that F1 score and Precision have around 77% which indicates that the model can precisely give result of churning data with 77% accuracy.